executor/join.go

```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
+		if outerResult.err != nil {
+			e.finished.Store(true)
+			return
+		}
+	}
+}
+
+// prepare4Chunk follows these steps:
+// step 1. fetch data from inner child and build a hash table;
+// step 2. fetch data from outer child in a background goroutine and probe the hash table in multiple join workers
+//
+// step 2 can be split into following work:
+//	+---------+-----------------------+--------------------------+
+//	|         |                       |                          |
+//	|         v                       v                          v
+//	|  outerChkResource[0]   outerChkResource[1] ...  outerChkResource[n]
+//	|         +                       +                  +
+//	|         |                       |                  |
+//	|         |                       v                  |
+//	|         |      +--+ background goroutine +--+      |
+//	|         +----> |                            | <----+
+//	|                |      fetchOuterChunks      |
+//	|            +---+                            +----------+
+//	|            |   +----------------+-----------+          |
+//	|            |                    |                      |
+//	|            v                    v                      v
+//	|      outerResultCh[0]    outerResultCh[1]  ...   outerResultCh[n]
+//	|            |                    |                      |
+//	+------------o----+---------------o----+-----------------o---+
+//	             |    |               |    |                 |   |
+//	             v    |               v    |                 v   |
+//	  +---goroutine 0---+   +---goroutine 1---+     +---goroutine n---+
+//	  |                 |   |                 |     |                 |
+//	+-+   joinWorker    | +-+   joinWorker    | ... |   joinWorker    +---+
+//	| |                 | | |                 |     |                 |   |
+//	| +------+----------+ | +--------+--------+     +---------+-------+   |
+//	|        ^            |          ^                        ^           |
+//	|        |            |          |                        |           |
+//	|        +            |          +                        +           |
+//	|  joinChkResource[0] |  joinChkResource[1] ...   joinChkResource[n]  |
+//	|        ^            |   ^                            ^              |
+//	|        |            |   |                            |              |
+//	+--------o------------+---o--->joinResultCh<-----------o--------------+
+//	         |                |       +                    |
+//	         |                |       |                    |
+//	         |                |       v                    |
+//	         +----------------+--+NextChunk+---------------+
+
+func (e *HashJoinExec) prepare4Chunk(goCtx goctx.Context) (err error) {
+	if err = e.build(goCtx); err != nil {
+		return errors.Trace(err)
+	}
+	e.probe(goCtx)
+	return nil
+}
+
+func (e *HashJoinExec) fetchInnerResults(goCtx goctx.Context) (err error) {
```

Comment:
1. how about `s/fetchInnerResults/fetchSelectedInners/` ?
2. add a comment, like: `fetchInnerResults fetchs the selected inner rows and copies them to "e.innerResult"`